flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/SessionReconciler.java

```diff
@@ -134,4 +142,31 @@ protected void shutdown(FlinkDeployment deployment) {
         flinkService.deleteClusterDeployment(
                 deployment.getMetadata(), deployment.getStatus(), true);
     }
+
+    @Override
+    public DeleteControl cleanup(FlinkDeployment flinkApp, Context context) {
+        var sessionJobs =
+                informerManager
+                        .getSessionJobInformer(flinkApp.getMetadata().getNamespace())
+                        .getIndexer()
+                        .byIndex(OperatorUtils.CLUSTER_ID_INDEX, flinkApp.getMetadata().getName());
+        if (!sessionJobs.isEmpty()) {
+            var error =
+                    String.format(
+                            "The session jobs %s should be deleted first",
+                            sessionJobs.stream()
+                                    .map(job -> job.getMetadata().getName())
+                                    .collect(Collectors.toList()));
+            // TODO generate error events for this
```

Comment:
I think events will be better here to help user know what happen. But, I have problem when generate the events here, I found that the example in `DeploymentFailedException#asEvent` can not work (It seems lack some of required parameters.)  I will improve this after figuring it out.